### PR TITLE
[compiler-rt] Work around incompatible Windows definitions of (S)SIZE_T

### PR DESCRIPTION
**Pull request by arichardson**
The interceptor types are supposed to match size_t (and the non-Windows
ssize_t) exactly, but on 32-bit Windows `size_t` uses `unsigned int`
whereas `SIZE_T` is `unsigned long`. The current definition results in
`uptr` not matching `uintptr_t` since we otherwise get typedef
redefinition errors. Work around this by using a #define instead of
a typedef when defining SIZE_T.

It would probably be cleaner to stop using these uppercase types, but
that is a rather invasive change and this one is the minimal change to
allow uptr to match uintptr_t on Windows.

To ensure this compiles on Windows, we also remove the interceptor.h
defines of uptr (that do not always match __sanitizer::uptr) and rely
on __sanitizer::uptr instead. The interceptor types most likely predate
those other types so clean up the unnecessary definition while here.

This also reverts commit 18e06e3e2f3d47433e1ed323b8725c76035fc1ac and
commit bb27dd853a713866c025a94ead8f03a1e25d1b6e.

### CODE REVIEW

File: compiler-rt/lib/interception/interception_type_test.cpp [hunk 1]
```diff
@@ -12,17 +12,26 @@
 //===----------------------------------------------------------------------===//
 
 #include "interception.h"
+#include "sanitizer_common/sanitizer_type_traits.h"
 
-#if SANITIZER_LINUX || SANITIZER_APPLE
```
Comment (arichardson) [inline compiler-rt/lib/interception/interception_type_test.cpp:16]: This change is unrelated, but IMO we should check these static asserts everywhere possible.

**Review by vitalybuka — APPROVED** [review]

**Review by mstorsjo — CHANGES_REQUESTED** [review]
Unfortunately, this doesn't work.

From the mingw build:
```
In file included from /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception_win.cpp:132:
In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/windows.h:69:
In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/windef.h:9:
In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/minwindef.h:163:
In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/winnt.h:150:
/home/martin/clang-trunk/i686-w64-mingw32/include/basetsd.h:147:39: error: typedef declarator cannot be qualified
  147 |   __MINGW_EXTENSION typedef ULONG_PTR SIZE_T,*PSIZE_T;
      |                                       ^~~~~~
/home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception.h:34:29: note: expanded from macro 'SIZE_T'
   34 | #define SIZE_T __sanitizer::usize
      |                ~~~~~~~~~~~~~^
```

From the MSVC build:
```
C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): error C2665: '__interception::OverrideFunction': no overloaded function could convert all the argument types
C:\code\llvm-project\compiler-rt\lib\interception\interception_win.h(31): note: could be 'bool __interception::OverrideFunction(const char *,__interception::uptr,__interception::uptr *)'
C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: 'bool __interception::OverrideFunction(const char *,__interception::uptr,__interception::uptr *)': cannot convert argument 3 from '__sanitizer::uptr *' to '__interception::uptr *'
C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(197): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or parenthesized function-style cast
C:\code\llvm-project\compiler-rt\lib\interception\interception_win.h(28): note: or       'bool __interception::OverrideFunction(__interception::uptr,__interception::uptr,__interception::uptr *)'
C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: 'bool __interception::OverrideFunction(__interception::uptr,__interception::uptr,__interception::uptr *)': cannot convert argument 1 from 'const char [15]' to '__interception::uptr'
C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: There is no context in which this conversion is possible
C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: while trying to match the argument list '(const char [15], __sanitizer::uptr, __sanitizer::uptr *)'
```

**Comment by arichardson** [comment]
> Unfortunately, this doesn't work.
> 
> From the mingw build:
> 
> ```
> In file included from /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception_win.cpp:132:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/windows.h:69:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/windef.h:9:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/minwindef.h:163:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/winnt.h:150:
> /home/martin/clang-trunk/i686-w64-mingw32/include/basetsd.h:147:39: error: typedef declarator cannot be qualified
>   147 |   __MINGW_EXTENSION typedef ULONG_PTR SIZE_T,*PSIZE_T;
>       |                                       ^~~~~~
> /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception.h:34:29: note: expanded from macro 'SIZE_T'
>    34 | #define SIZE_T __sanitizer::usize
>       |                ~~~~~~~~~~~~~^
> ```
> 
> From the MSVC build:
> 
> ```
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): error C2665: '__interception::OverrideFunction': no overloaded function could convert all the argument types
> C:\code\llvm-project\compiler-rt\lib\interception\interception_win.h(31): note: could be 'bool __interception::OverrideFunction(const char *,__interception::uptr,__interception::uptr *)'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: 'bool __interception::OverrideFunction(const char *,__interception::uptr,__interception::uptr *)': cannot convert argument 3 from '__sanitizer::uptr *' to '__interception::uptr *'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(197): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or parenthesized function-style cast
> C:\code\llvm-project\compiler-rt\lib\interception\interception_win.h(28): note: or       'bool __interception::OverrideFunction(__interception::uptr,__interception::uptr,__interception::uptr *)'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: 'bool __interception::OverrideFunction(__interception::uptr,__interception::uptr,__interception::uptr *)': cannot convert argument 1 from 'const char [15]' to '__interception::uptr'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: There is no context in which this conversion is possible
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: while trying to match the argument list '(const char [15], __sanitizer::uptr, __sanitizer::uptr *)'
> ```

Thanks for testing this - I will try to rework it to fix the problem.

**Comment by arichardson** [comment]
> Unfortunately, this doesn't work.
> 
> From the mingw build:
> 
> ```
> In file included from /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception_win.cpp:132:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/windows.h:69:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/windef.h:9:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/minwindef.h:163:
> In file included from /home/martin/clang-trunk/i686-w64-mingw32/include/winnt.h:150:
> /home/martin/clang-trunk/i686-w64-mingw32/include/basetsd.h:147:39: error: typedef declarator cannot be qualified
>   147 |   __MINGW_EXTENSION typedef ULONG_PTR SIZE_T,*PSIZE_T;
>       |                                       ^~~~~~
> /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception.h:34:29: note: expanded from macro 'SIZE_T'
>    34 | #define SIZE_T __sanitizer::usize
>       |                ~~~~~~~~~~~~~^
> ```
> 
> From the MSVC build:
> 
> ```
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): error C2665: '__interception::OverrideFunction': no overloaded function could convert all the argument types
> C:\code\llvm-project\compiler-rt\lib\interception\interception_win.h(31): note: could be 'bool __interception::OverrideFunction(const char *,__interception::uptr,__interception::uptr *)'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: 'bool __interception::OverrideFunction(const char *,__interception::uptr,__interception::uptr *)': cannot convert argument 3 from '__sanitizer::uptr *' to '__interception::uptr *'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(197): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or parenthesized function-style cast
> C:\code\llvm-project\compiler-rt\lib\interception\interception_win.h(28): note: or       'bool __interception::OverrideFunction(__interception::uptr,__interception::uptr,__interception::uptr *)'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: 'bool __interception::OverrideFunction(__interception::uptr,__interception::uptr,__interception::uptr *)': cannot convert argument 1 from 'const char [15]' to '__interception::uptr'
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: There is no context in which this conversion is possible
> C:\code\llvm-project\compiler-rt\lib\asan\asan_win.cpp(195): note: while trying to match the argument list '(const char [15], __sanitizer::uptr, __sanitizer::uptr *)'
> ```

I tried setting up a Windows VM, but was not able to get this working today. However, I managed to reproduce these build failures with mingw packages and:
```
-DCMAKE_C_COMPILER=/usr/bin/clang
-DCMAKE_CXX_COMPILER=/usr/bin/clang++
-DCOMPILER_RT_INCLUDE_TESTS=ON
-DCOMPILER_RT_DEBUG=OFF
-DCOMPILER_RT_BUILD_PROFILE=ON
-DCOMPILER_RT_BUILD_XRAY=OFF
-DCOMPILER_RT_BUILD_SANITIZERS=ON
-DCOMPILER_RT_BUILD_LIBFUZZER=OFF
-DCOMPILER_RT_BUILD_GWP_ASAN=ON
-DCOMPILER_RT_BUILD_ORC=OFF
-DCOMPILER_RT_BUILD_MEMPROF=ON
-DCOMPILER_RT_BUILD_CRT=OFF
-DCOMPILER_RT_BUILD_BUILTINS=ON
-DLLVM_CMAKE_DIR=../../build/upstream-llvm-project-build
-DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON
-DCMAKE_C_COMPILER_TARGET=i686-w64-windows-gnu
-DCMAKE_CXX_COMPILER_TARGET=i686-w64-windows-gnu
-DCMAKE_SYSTEM_NAME=Windows
-DCOMPILER_RT_CAN_EXECUTE_TESTS=ON
-DCMAKE_CROSSCOMPILING_EMULATOR=/usr/bin/wine
-DCMAKE_EXE_LINKER_FLAGS_INIT=-L/usr/lib/gcc/i686-w64-mingw32/13-win32/
-DCMAKE_SHARED_LINKER_FLAGS_INIT=-L/usr/lib/gcc/i686-w64-mingw32/13-win32/
```

I can't successfully build the libraries since I get linker errors, but I've confirmed that the latest version of this patch fixes the compilation errors reported by @mstorsjo

**Comment by mstorsjo** [comment]
> I can't successfully build the libraries since I get linker errors, but I've confirmed that the latest version of this patch fixes the compilation errors reported by @mstorsjo

Indeed, this version does build successfully with MSVC, and it does get past the earlier quoted errors when building for mingw targets, but for mingw x86_64 (and MSVC x86_64 too), it fails on other issues due to some of the other, new static asserts you've added here:
```
/home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception_type_test.cpp:44:16: error: static assertion failed due to requirement 'sizeof(unsigned long long) == sizeof(long)': 
   44 | COMPILER_CHECK(sizeof(::OFF_T) == sizeof(off_t));
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/../sanitizer_common/sanitizer_internal_defs.h:357:44: note: expanded from macro 'COMPILER_CHECK'
  357 | #define COMPILER_CHECK(pred) static_assert(pred, "")
      |                                            ^~~~
/home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception_type_test.cpp:44:32: note: expression evaluates to '8 == 4'
   44 | COMPILER_CHECK(sizeof(::OFF_T) == sizeof(off_t));
      |                ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/../sanitizer_common/sanitizer_internal_defs.h:357:44: note: expanded from macro 'COMPILER_CHECK'
  357 | #define COMPILER_CHECK(pred) static_assert(pred, "")
      |                                            ^~~~
1 error generated.
```

FWIW, it should be quite easy to test the build for mingw targets, with llvm-mingw. To do that:

1. Clone https://github.com/mstorsjo/llvm-mingw
2. Within your `llvm-mingw` checkout, check out (or symlink) your `llvm-project` repo, so that you have `llvm-mingw/llvm-project` containing your compiler-rt code to be tested
3. Grab an existing release, e.g. https://github.com/mstorsjo/llvm-mingw/releases/tag/20240820, and download a suitable package for you (e.g. an x86_64 linux build), e.g. [llvm-mingw-20240820-ucrt-ubuntu-20.04-x86_64.tar.xz](https://github.com/mstorsjo/llvm-mingw/releases/download/20240820/llvm-mingw-20240820-ucrt-ubuntu-20.04-x86_64.tar.xz), and unpack that somewhere
4. Within the `llvm-mingw` directory, run `./build-compiler-rt.sh --build-sanitizers <path-to-unpacked-toolchain>`. That uses the toolchain from step 3, which is expected to be unpacked in `<path-to-unpacked-toolchain>`, building the compiler-rt code in `llvm-mingw/llvm-project` and installing it on top of your unpacked toolchain. This tries building it for both `i686` and `x86_64`.

The compiler-rt sanitizers are buildable for armv7 and aarch64 too, but the `build-compiler-rt.sh` script skips that for now (it's not buildable for armv7 in the 19.x release branch), but if you want to, you can rip out the code for skipping those arches in the script, in order to test building for all 4 architectures.

**Comment by arichardson** [comment]
> > I can't successfully build the libraries since I get linker errors, but I've confirmed that the latest version of this patch fixes the compilation errors reported by @mstorsjo
> 
> Indeed, this version does build successfully with MSVC, and it does get past the earlier quoted errors when building for mingw targets, but for mingw x86_64 (and MSVC x86_64 too), it fails on other issues due to some of the other, new static asserts you've added here:
> 
> ```
> /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception_type_test.cpp:44:16: error: static assertion failed due to requirement 'sizeof(unsigned long long) == sizeof(long)': 
>    44 | COMPILER_CHECK(sizeof(::OFF_T) == sizeof(off_t));
>       |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/../sanitizer_common/sanitizer_internal_defs.h:357:44: note: expanded from macro 'COMPILER_CHECK'
>   357 | #define COMPILER_CHECK(pred) static_assert(pred, "")
>       |                                            ^~~~
> /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/interception_type_test.cpp:44:32: note: expression evaluates to '8 == 4'
>    44 | COMPILER_CHECK(sizeof(::OFF_T) == sizeof(off_t));
>       |                ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
> /home/martin/code/llvm-mingw/llvm-project/compiler-rt/lib/interception/../sanitizer_common/sanitizer_internal_defs.h:357:44: note: expanded from macro 'COMPILER_CHECK'
>   357 | #define COMPILER_CHECK(pred) static_assert(pred, "")
>       |                                            ^~~~
> 1 error generated.
> ```
> 
> FWIW, it should be quite easy to test the build for mingw targets, with llvm-mingw. To do that:
> 
> 1. Clone https://github.com/mstorsjo/llvm-mingw
> 2. Within your `llvm-mingw` checkout, check out (or symlink) your `llvm-project` repo, so that you have `llvm-mingw/llvm-project` containing your compiler-rt code to be tested
> 3. Grab an existing release, e.g. https://github.com/mstorsjo/llvm-mingw/releases/tag/20240820, and download a suitable package for you (e.g. an x86_64 linux build), e.g. [llvm-mingw-20240820-ucrt-ubuntu-20.04-x86_64.tar.xz](https://github.com/mstorsjo/llvm-mingw/releases/download/20240820/llvm-mingw-20240820-ucrt-ubuntu-20.04-x86_64.tar.xz), and unpack that somewhere
> 4. Within the `llvm-mingw` directory, run `./build-compiler-rt.sh --build-sanitizers <path-to-unpacked-toolchain>`. That uses the toolchain from step 3, which is expected to be unpacked in `<path-to-unpacked-toolchain>`, building the compiler-rt code in `llvm-mingw/llvm-project` and installing it on top of your unpacked toolchain. This tries building it for both `i686` and `x86_64`.
> 
> The compiler-rt sanitizers are buildable for armv7 and aarch64 too, but the `build-compiler-rt.sh` script skips that for now (it's not buildable for armv7 in the 19.x release branch), but if you want to, you can rip out the code for skipping those arches in the script, in order to test building for all 4 architectures.

Thank you very much for those instructions! I have now built for Win64 and avoided the static_assert. I plan to clean up off_t handling in a separate PR.

**Review by mstorsjo — APPROVED** [review]
LGTM, this seems to work fine now. (I don't have a strong opinion on whether this is a good way of solving it or not though.)

**Comment by wrotki** [comment]
We're seeing a build failures on Mac LLVM compile like:
....
.../llvm-project/compiler-rt/lib/interception/interception_type_test.cpp:26:17: error: static assertion failed due to requirement '__sanitizer::is_same<long, int>::value': 
   26 | COMPILER_CHECK((__sanitizer::is_same<::PTRDIFF_T, ::ptrdiff_t>::value));
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.../llvm-project/compiler-rt/lib/interception/../sanitizer_common/sanitizer_internal_defs.h:362:44: note: expanded from macro 'COMPILER_CHECK'
  362 | #define COMPILER_CHECK(pred) static_assert(pred, "")
      |                                            ^~~~
1 error generated.
....

I'm trying to reproduce it locally - but until I do and figure out why exactly, would you consider reverting? Thanks

**Comment by arichardson** [comment]
> We're seeing a build failures on Mac LLVM compile like: .... .../llvm-project/compiler-rt/lib/interception/interception_type_test.cpp:26:17: error: static assertion failed due to requirement '__sanitizer::is_same<long, int>::value': 26 | COMPILER_CHECK((__sanitizer::is_same<::PTRDIFF_T, ::ptrdiff_t>::value)); | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ .../llvm-project/compiler-rt/lib/interception/../sanitizer_common/sanitizer_internal_defs.h:362:44: note: expanded from macro 'COMPILER_CHECK' 362 | #define COMPILER_CHECK(pred) static_assert(pred, "") | ^~~~ 1 error generated. ....
> 
> I'm trying to reproduce it locally - but until I do and figure out why exactly it is happening in our configuration, would you consider reverting? Thanks

It sounds like on macos 32 bit ptrdiff_t is int rather than long. You could replace `COMPILER_CHECK((__sanitizer::is_same<::PTRDIFF_T, ::ptrdiff_t>::value));` with the previous `COMPILER_CHECK(sizeof(::PTRDIFF_T) == sizeof(ptrdiff_t));`. The real fix would be to define `PTRDIFF_T` as int for your 32-bit configuration.

Can you tell me which target you are building for?

**Comment by arichardson** [comment]
The other option would be 

```
#if SANITIZER_APPLE && SANITIZER_WORDSIZE == 32
typedef int    PTRDIFF_T;
#else
typedef __sanitizer::sptr    PTRDIFF_T;
#endif
```

**Comment by wrotki** [comment]
My clang command line has  `-arch arm64 -arch i386 -arch x86_64` , while I only see `-arch arm64 -arch x86_64` in open source build, which is probably why it works there, only 64 bit.

Let me experiment with your suggestions, thanks!

**Pull request merged**